pkg/capacity/capacity.go

```diff
@@ -187,8 +187,6 @@ func NewCentralCapacityController(
 func (c *Controller) Run(ctx context.Context, threadiness int) {
 	klog.Info("Starting Capacity Controller")
 	defer c.queue.ShutDown()
-	go c.scInformer.Informer().Run(ctx.Done())
-	go c.topologyInformer.Run(ctx)
```

Comment:
There is at least one other redundant Run:
https://github.com/kubernetes-csi/external-provisioner/blob/bcf1a405b5f6a0fa2aadb80d08c9c2569c4b6912/pkg/controller/clone_controller.go#L82

On the other hand, this code does avoid Run if the informer is passed in:
https://github.com/kubernetes-csi/external-provisioner/blob/bcf1a405b5f6a0fa2aadb80d08c9c2569c4b6912/vendor/sigs.k8s.io/sig-storage-lib-external-provisioner/v6/controller/controller.go#L863-L873

What happens when `Run` is called too often? I.e., what goes wrong?

Reply:
I think all these places should be fixed. I'm not sure how often it will cause problem. And In my case, the redundant CSIStorageCapacities is not showed every time. But I check the source code. In informer.run(), it will call controller.run(), and the comment said it 's an error to call Run more than once

// Run begins processing items, and will continue until a value is sent down stopCh.
// It's an error to call Run more than once.
// Run blocks; call via go.
func (c *controller) Run(stopCh <-chan struct{}) {
	defer utilruntime.HandleCrash()
	go func() {
		<-stopCh
		c.config.Queue.Close()
	}()
	r := NewReflector(
		c.config.ListerWatcher,
		c.config.ObjectType,
		c.config.Queue,
		c.config.FullResyncPeriod,
	)
	r.ShouldResync = c.config.ShouldResync
	r.clock = c.clock

	c.reflectorMutex.Lock()
	c.reflector = r
	c.reflectorMutex.Unlock()

	var wg wait.Group
	defer wg.Wait()

	wg.StartWithChannel(stopCh, r.Run)

	wait.Until(c.processLoop, time.Second, stopCh)
}

Reply:
BTW, after I remove these two lines, the redundant CSIStorageCapacities case is not happened again. Now I try to run this test case for one or two days to check whether the case happens again

Reply:
So the negative effect was that the capacity controller created redundant objects, probably because its informer callbacks were called too often. I wonder whether the duplicates should have been avoided in this case. This might be worth investigating, just to get full closure on the issue.

The fix itself here makes sense.

/kind bug

Can you add a change note that explains the bug? Instead of "NONE" in the description, put something like "Redundant CSIStorageCapacity objects might have been created because informers were not handled correctly."

Reply:
OK, no problem